### DISCUSSION TITLE
Can any user retrieve the s3 secret of another user before him ?

### DISCUSSION THREAD

**Discussion by azro352** [apache/ozone#6637] [FAQ] [answered]
Hi

I was reading some parts of the code and found an undocumented endpoint `PUT /secret/{username}` that allows to retrieve a specific user credentials

As I don't see any validation or else, does that mean that a any user **A** can ask for the credentials of a user **B**, and 

- if user **B** never asked for his credentials, the **A** get them, and **B** won't be able to retrieve, he'd need to revoke and generate
- if user **B** already asked for his credentials, then **A** gets an error ? 

Does that means that this endpoint shouldn't be directly accessible by users ? Like endpoint `DELETE /secret/{username}`, if you can revoke another user credential, that could cause trouble

Thanks

**Comment by kerneltime** [comment]
@azro352 you can find more details for the change here https://issues.apache.org/jira/browse/HDDS-8050
This is protected by kerberos and is a way for users to get their own S3 credentials. 
Did you encounter a problem with this feature?
The relevant tests/sample usage
[hadoop-ozone/dist/src/main/smoketest/s3/secretgenerate.robot](https://github.com/apache/ozone/blob/master/hadoop-ozone/dist/src/main/smoketest/s3/secretgenerate.robot)
and 
[hadoop-ozone/dist/src/main/smoketest/s3/secretrevoke.robot](https://github.com/apache/ozone/blob/master/hadoop-ozone/dist/src/main/smoketest/s3/secretrevoke.robot)

**Reply by azro352** [reply]
No the code that I'm talking about comes from https://issues.apache.org/jira/browse/HDDS-9203, when endpointq for creating and revoking secrets for a specific user (not taking the username from the kerberos context) were added


The link you show confirms my initial supposition with the test : `S3 Gateway Generate Secret By Username For Other User`

https://github.com/apache/ozone/blob/master/hadoop-ozone/dist/src/main/smoketest/s3/secretgenerate.robot#L53

**Reply by kerneltime** [reply]
The kerberos identity calling the endpoint should be a registered S3 admin `ozone.s3.administrators`. This allows the administrator of S3 to create credentials for other users.  Ref: `hadoop-hdds/docs/content/security/SecuringS3.md` and `hadoop-ozone/dist/src/main/smoketest/s3/secretgenerate.robot`

**Reply by kerneltime** [reply]
An administrator can invalidate or delete credentials. A random user cannot step on another user.

**Reply by kerneltime** [reply]
If you would like feel free to contribute a robot test for the scenarios you have in mind.

**Reply by azro352** [reply]
I don't see any mention of  `admin` in the 2 links you gave, neither the test nor the documentation

Also I don't see any difference of validation between

- [`PUT /secret`](https://github.com/apache/ozone/blob/master/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretManagementEndpoint.java#L48) each user retrieve its secret
- [`PUT /secret/<username>`](https://github.com/apache/ozone/blob/master/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretManagementEndpoint.java#L53) rerieve secret of other user

**Reply by ChenSammi** [reply, marked as answer]
The ozone administrator or same user validation is checked at server side(OM), not client side(s3 gateway)
https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/security/S3SecretRequestHelper.java#L68

**Comment by errose28** [comment]
@ChenSammi is correct, validation is present within the OM, not on the S3 gateway. I'm currently closing this discussion since security should be discussed on [security@ozone.apache.org](mailto:security@ozone.apache.org) per this repository (and the ASF's) [security policy](https://github.com/apache/ozone/blob/master/SECURITY.md).
